Terrain protocol docs

hamishwillee opened the pull request:
@DonLakeFlyer Probably good to capture this. 

This is a placeholder with questions - I don't understand the sequence and what happens with timeouts. For example, even who kicks off the conversation, and whether it can be used outside missions. 

A good place to start would be proper sequences showing timeouts. That is now easy you can [click the link here for online tool](https://mermaid-js.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoic2VxdWVuY2VEaWFncmFtO1xuICAgIHBhcnRpY2lwYW50IEdDU1xuICAgIHBhcnRpY2lwYW50IERyb25lXG4gICAgRHJvbmUtPj5HQ1M6IFRFUlJBSU5fUkVRVUVTVFxuICAgIERyb25lLT4-RHJvbmU6IFN0YXJ0IHRpbWVvdXRcbiAgICBHQ1MtPj5Ecm9uZTogVEVSUkFJTl9EQVRBXG4gICAgTm90ZSByaWdodCBvZiBHQ1M6IFNhbWUgb3IgZGlmZmVyZW50IHNlcXVlbmNlPz8_P1xuICAgIEdDUy0-PkRyb25lOiBURVJSQUlOX0NIRUNLXG4gICAgRHJvbmUtPj5HQ1M6IFRFUlJBSU5fUkVQT1JUIiwibWVybWFpZCI6eyJ0aGVtZSI6ImRlZmF1bHQifSwidXBkYXRlRWRpdG9yIjpmYWxzZX0) - fix up, click the copy markdown link, and paste back.

Or we could have  brief slack chat if that helps?

Anyway, the questions are in the placeholder content.

hamishwillee commented:
Also see https://github.com/mavlink/mavlink/issues/1314

> Another thing which wasn't clear to me from reading spec was on TERRAIN_REQUEST.mask. Bit 0 position (TERRAIN_REQUEST.mask = 1) in that mask refers to the last row first column in the 8x7 array of grids. I initially assumed it was first row/column.


AND

Spec says:

- pending - Number of 4x4 terrain blocks waiting to be received or read from disk
- loaded - Number of 4x4 terrain blocks in memory

But from the work i'm doing to support in QGC it seems like this is really:

- pending - number of blocks the vehicle is still waiting on from the gcs
- loaded - number of blocks the vehicle currently has

> From Randy I think the two are pretty much the same, it's just that spec's version doesnt' mention that it's speaking from the vehicle/autopilot's point of view instead of from the GCSs point of view.

DonLakeFlyer commented:
This is as much as I know about the tile protocol stuff:

TERRAIN_DATA | - | - | -
--- | --- | --- | ---
lat	| int32_t | degE7 |	Latitude of SW corner of 4x4 tile
lon	| int32_t	| degE7 | Longitude of SW corner of 4x4 tile
grid_spacing | uint16_t | m | Distance between each terrain height in the 4x4 tile in a north/south and east/west direction
gridbit |	uint8_t | | Bit within the `TERRAIN_REQUEST.mask` for which this data represents
data | int16_t[16] | m	| 4x4 tile of MSL terrain data

* Use by GCS to send a 4x4 tile of AMSL heights to the Vehicle
* Sent by a GCS in response to a `TERRAIN_REQUEST`

TERRAIN_REQUEST | - | - | -
--- | --- | --- | ---
lat | int32_t |	degE7 | Latitude of SW corner of first 4x4 block
lon	| int32_t	| degE7 | Longitude of SW corner of first 4x4 block
grid_spacing | uint16_t | m | Distance between each terrain height in the 4x4 grid in a north/south and east/west direction
mask | uint64_t	| | I explained below how it works

* Sent by Vehicle to request terrain tile(s) to be uploaded
* GCS responds to this by sending a TERRAIN_DATA messages to the Vehicle for each bit set in `TERRAIN_REQUEST.mask`
* There is no back/forth protocol or retries or anything like that from the GCS side
* The Vehicle side is fully in control of re-requesting whatever data it needs by sending TERRAIN_REQUEST again for anything missing
* The bits within `mask` refer to an 8 columns x 7 rows grid
* Each spot in the 8x7 grid refers to a specific 4x4 tile
* Lowest order bit in `TERRAIN_REQUEST.mask` refers to the SW corner of the 8x7 grid
* Bits then proceed east across the 8 columns in the current row
* When the columns in a row are exhausted the next bit indicates a move north on the grid and back to the west edge

TERRAIN_REPORT | - | - | -
--- | --- | --- | ---
lat	| int32_t | degE7 | Vehicle position latitude
lon	| int32_t | degE7 | Vehicle position longitude
spacing |	uint16_t | m | grid spacing (zero if terrain at this location unavailable)
terrain_height |	float | m	| Terrain height MSL at lat/lon known to Vehicle
current_height |	float | m	| I don't use this, so I don't know what it really is
pending |	uint16_t | |	Number of 4x4 terrain blocks Vehicle is waiting on from GCS
loaded	| uint16_t | | Number of 4x4 terrain blocks currently loaded onto Vehicle

* Sent by Vehicle to GCS
* Used by QGC to show tile upload progress indicator
* Once pending == 0, then theoretically the Vehicle has all the information it needs to fly
* ArduPilot continuously streams this to the GCS

hamishwillee commented:
Hi @DonLakeFlyer,
That's helpful thank you.
1. Are bits numbered from 0 or 1? i.e. Is `TERRAIN_DATA.gridbit` 0-55 or 1-56?
2. So this is how the data looks?
    ![image](https://user-images.githubusercontent.com/5368500/86999311-a786d080-c1f5-11ea-835a-d3e746c0c141.png)
3. The docs make it pretty clear that [TERRAIN_CHECK](https://mavlink.io/en/messages/common.html#TERRAIN_CHECK) is intended to request `TERRAIN_REPORT`, which in turn makes it clear what the `TERRAIN_REPORT.current_height` was intended for. I guess the theory was at some point that the GCS might check what the flight stack needs.


Regards
H

DonLakeFlyer commented:
* `TERRAIN_DATA.gridbit` is a 0-based.
* Your image is correct
* With ArduPilot `TERRAIN_REPORT` is constantly streamed. You don't need to ever ask for it. 
* I don't see how this comment in the spec on `TERRAIN_CHECK` "Used by GCS to check if vehicle has all terrain data needed for a mission." makes sense. The only way for a GCS to know when a vehicle has all the terrain info it needs to fly is when `TERRAIN_REPORT.pending` = 0.
* You'll have to ask the ArduPilot folks what `TERRAIN_REPORT.current_height` is. It's a vehicle height (according to spec) but what kind: relative to home, relative to terrain, AMSL?

bkueng commented:
I implemented this on PX4, and this looks good. I like the diagram with the indexes.

One thing I got wrong from the comment above (though it's correct in the diff): `TERRAIN_DATA:` `lat 	int32_t 	degE7 	Latitude of SW corner of 4x4 tile`. It should be the lat of the `TERRAIN_REQUEST` not the 4x4 tile.

I did not implement `TERRAIN_CHECK`, as `TERRAIN_REPORT` is streamed already.
Also for the report, I do not fill out the lat, lon or height. I don't see the use for that, as there are other messages for that already.

DonLakeFlyer commented:
> Also for the report, I do not fill out the lat, lon or height. I don't see the use for that, as there are other messages for that already.

That is useful for a GCS to check its own internal terrain data against what the vehicle thinks it is. I used it quite a bit when debugging QGC code (lat/lon/terrain_height).

hamishwillee commented:
@bkueng @DonLakeFlyer That was useful 
- I had misinterpreted the lat, lon too. Fixed up in the image.
- I also removed the TERRAIN_CHECK from sequences, explicitly noting it is not part of the protocol. I will deprecate it in the XML too. 
- Added note as to why including the lat, lon, height in terrain report is useful.

Merged, as it is IMO good now.

DonLakeFlyer commented:
> * I also removed the TERRAIN_CHECK from sequences, explicitly noting it is not part of the protocol. I will deprecate it in the XML too.

YOu should ask ArduPilot folks first if they support that and if so what it does

hamishwillee commented:
OK. I sent an email to https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!forum/ardupilot-gcs

mnumanuyar commented:
Will there be a way to initiate Terrain data transfer from gcs side ?  I think  to be able to upload terrain info from gcs to vehicle would be usefull (for example if the user thinks vehicle might lost connection near a dangerous terrain).  Otherwise user had to upload this info to sd card on vehicle, which is not always easy to access.

hamishwillee commented:
@mnumanuyar No, the protocol is as described. The GCS does trigger the upload though - simply by specifying a mission or other flight that uses terrain data. There is no process for offline terrain data upload as far as I am aware. If you think this is useful for your particular use case you might come along to the MAVLink devcall and explain your use case and suggested solutions.

mnumanuyar commented:
@hamishwillee Ok thanks, I will try to join this wednesday, I also opened an issue about it [here ](https://github.com/mavlink/mavlink/issues/1459)

DonLakeFlyer commented:
I would also be interesting to understand if ArduPilot + MP already support this sort of thing somehow over the existing protocol. I don't know the answer to that.

hamishwillee commented:
@mnumanuyar Thanks. Added comments to your other issue. Maybe we can resolve before the meeting. But if not, added to the agenda: https://github.com/mavlink/mavlink/wiki/20200902-Dev-Meeting

DonLakeFlyer commented:
And the discussion here: https://discuss.ardupilot.org/t/terrain-generation-issue/60920 would seem to indicate user initiated upload to the vehicle is possible.

hamishwillee commented:
Not necessarily. It might be just "copy file into SD card in particular location". But hopefully. I asked here https://discuss.ardupilot.org/t/terrain-data-mavlink-protocol-and-offline-use/60942
